src/index/ivf_raft/ivf_raft.cuh

```diff
@@ -0,0 +1,385 @@
+                            }
+                        } catch (std::exception& e) {
+                            LOG_KNOWHERE_WARNING_ << "RAFT inner error, " << e.what();
+                            result = Status::raft_inner_error;
+                        }
+                    }
+
+                    return result;
+                }
+
+            virtual expected<DataSetPtr, Status>
+                Search(const DataSet& dataset, const Config& cfg, const BitsetView& bitset) const override {
+                    auto ivf_raft_cfg = static_cast<const typename KnowhereConfigType<T>::Type&>(cfg);
+                    auto rows = dataset.GetRows();
+                    auto dim = dataset.GetDim();
+                    auto* data = reinterpret_cast<float const*>(dataset.GetTensor());
+                    auto output_size = rows * ivf_raft_cfg.k;
+                    auto ids = std::unique_ptr<std::int64_t[]>(new std::int64_t[output_size]);
+                    auto dis = std::unique_ptr<float[]>(new float[output_size]);
+
+                    try {
+
+                        auto stream = res_->get_stream();
+                        auto data_gpu = rmm::device_uvector<float>(rows * dim, stream);
+                        RAFT_CUDA_TRY(
+                                cudaMemcpyAsync(
+                                    data_gpu.data(),
+                                    data,
+                                    data_gpu.size() * sizeof(float),
+                                    cudaMemcpyDefault,
+                                    stream.value()
+                                    )
+                                );
+
+                        auto ids_gpu = rmm::device_uvector<std::int64_t>(output_size, stream);
+                        auto dis_gpu = rmm::device_uvector<float>(output_size, stream);
+
+                        if constexpr (std::is_same_v<detail::raft_ivf_flat_index, T>) {
+                            auto search_params = raft::neighbors::ivf_flat::search_params{};
+                            search_params.n_probes = ivf_raft_cfg.nprobe;
+                            raft::neighbors::ivf_flat::search<float, std::int64_t>(
+                                    *res_,
+                                    search_params,
+                                    *gpu_index_,
+                                    data_gpu.data(),
+                                    rows,
+                                    ivf_raft_cfg.k,
+                                    ids_gpu.data(),
+                                    dis_gpu.data()
+                                    );
+                        } else if constexpr (std::is_same_v<detail::raft_ivf_pq_index, T>) {
+                            auto search_params = raft::neighbors::ivf_pq::search_params{};
+                            search_params.n_probes = ivf_raft_cfg.nprobe;
+                            raft::neighbors::ivf_pq::search<float, std::int64_t>(
```

Comment:
Just a heads up, the pointer-based APIs are going to be going away very soon. I'd consider using the mdspan API for IVF Flat since it's available now. We have a PR to add the mdspan API for IVF PQ going in soon. We might be removing the pointer APIs as soon as 23.04.